src/3d/qgstessellator.cpp

```diff
@@ -157,68 +182,160 @@ void QgsTessellator::addPolygon( const QgsPolygonV2 &polygon, float extrusionHei
-        holePolyline.push_back( pt2 );
-        float zPt = qIsNaN( pt.z() ) ? 0 : pt.z();
-        z[pt2] = zPt;
+        continue;
       }
-      ptPrev = pt;
+
+      p2t::Point *pt2 = new p2t::Point( x, y );
+      polyline.push_back( pt2 );
+
+      z[pt2] = qIsNaN( pt.z() ) ? 0 : pt.z();
     }
-    cdt->AddHole( holePolyline );
-    polylinesToDelete << holePolyline;
-  }
+    polylinesToDelete << polyline;
+
+    if ( polyline.size() < 3 )
+      return;
 
-  // TODO: robustness (no nearly duplicate points, invalid geometries ...)
+    p2t::CDT *cdt = new p2t::CDT( polyline );
+
+    // polygon holes
+    for ( int i = 0; i < polygon.numInteriorRings(); ++i )
+    {
+      std::vector<p2t::Point *> holePolyline;
+      holePolyline.reserve( exterior->numPoints() );
+      const QgsCurve *hole = polygon.interiorRing( i );
+      for ( int j = 0; j < hole->numPoints() - 1; ++j )
+      {
+        hole->pointAt( j, pt, vt );
+        QVector3D tempPt( pt.x(), pt.y(), ( qIsNaN( pt.z() ) ? 0 : pt.z() ) );
 
-  cdt->Triangulate();
+        const float x = QVector3D::dotProduct( tempPt - pOrigin, pXVector );
+        const float y = QVector3D::dotProduct( tempPt - pOrigin, pYVector );
 
-  std::vector<p2t::Triangle *> triangles = cdt->GetTriangles();
+        const bool found = std::find_if( polyline.begin(), polyline.end(), [x, y]( p2t::Point *&p ) { return *p == p2t::Point( x, y ); } ) != polyline.end();
 
-  for ( size_t i = 0; i < triangles.size(); ++i )
-  {
-    p2t::Triangle *t = triangles[i];
-    for ( int j = 0; j < 3; ++j )
+        if ( found )
+        {
+          continue;
+        }
+
+        p2t::Point *pt2 = new p2t::Point( x, y );
+        polyline.push_back( pt2 );
+
+        z[pt2] = qIsNaN( pt.z() ) ? 0 : pt.z();
+      }
+      cdt->AddHole( holePolyline );
+      polylinesToDelete << holePolyline;
+    }
+
+    // TODO: robustness (no nearly duplicate points, invalid geometries ...)
+
+    if ( polyline.size() == 3 && polygon.numInteriorRings() == 0 )
     {
-      p2t::Point *p = t->GetPoint( j );
-      float zPt = z[p];
-      mData << p->x << extrusionHeight + zPt << -p->y;
-      if ( mAddNormals )
-        mData << 0.f << 1.f << 0.f;
+      for ( std::vector<p2t::Point *>::iterator it = polyline.begin(); it != polyline.end(); it++ )
+      {
+        p2t::Point *p = *it;
+        const double zPt = z[p];
+        QVector3D nPoint = pOrigin + pXVector * p->x + pYVector * p->y;
+        const double fx = nPoint.x() - mOriginX;
+        const double fy = nPoint.y() - mOriginY;
+        const double fz = extrusionHeight + ( qIsNaN( zPt ) ? 0 : zPt );
+        mData << fx << fz << -fy;
+        if ( mAddNormals )
+          mData << pNormal.x() << pNormal.z() << - pNormal.y();
+      }
     }
-  }
+    else
+    {
+      try
+      {
+        cdt->Triangulate();
+      }
+      catch ( ... )
+      {
+        qDebug() << "Triangulation failed. Skipping polygon...";
+        return;
```

Comment:
if I understand well, there is a leak of polyline points if we return here.

replacing `std::vector<p2t::Point *> polyline` by `std::vector<std::unique_prt< p2t::Point > > polyline` will remove worries of this sort. But I see that the CDT interface requires plain pointers. 

I think polylinesToDelete should be a custom (local) class to handle resources, and both polyline and holes ownership should be transfered to this class right at the moment of creation.

```c++
struct PointVectorHandler:
     ~PointVectorHandler() 
    {
      for ( int i = 0; i < mPolylinesToDelete.count(); ++i )
          qDeleteAll( mPolylinesToDelete[i] )
    }

   std::vector< cdt::Point * > > *add( std::vector< cdt::Point * > > & v )
   {
     mPolylinesToDelete << v;
      return v;
    } 
private:
  std::vector< std::vector< cdt::Point * > > mPolyLinesToDelete;
```

also a leak of cdt wich sould be as std::unique_ptr.